libraries/Wire/examples/i2c-P2P-slave/i2c-P2P-slave.ino

```diff
@@ -0,0 +1,278 @@
+
+  Wire.onReceive(receiveEvent);
+
+  Serial.print("I2C Slave started on address: 0x0");
+  Serial.println(I2C_ADDRESS, HEX);
+}
+
+// should be in shared header
+MESSAGE_DATA encodeMessage(String bytes) {
+
+  msgdata.terminator = '.';
+
+  unsigned int datasize = sizeof(msgdata.data);
+  Serial.print("Encoding data of size: ");
+  Serial.println(datasize);
+
+  // Generate new data set for the struct
+  for (size_t i = 0; i < datasize; i++) {
+    if (bytes.length() >= i) {
+      msgdata.data[i] = bytes[i];
+    } else {
+      msgdata.data[i] = '\0';
+    }
+    Serial.print(msgdata.data[i]);
+    Serial.print(" ");
+  }
+  Serial.println();
+  msgdata.crc16 = calculateCRC16((uint8_t*) &msgdata.data[0], datasize);
+  Serial.print("Outgoing Message CRC16: "); Serial.println(msgdata.crc16, HEX);
+  return msgdata;
+}
+
+// should be in shared header
+MESSAGE_DATA validateMessage(char* bytes_message) {
+
+  MESSAGE_DATA tmp;
+  memcpy(&tmp, bytes_message, sizeof(tmp));
+
+  // Validate terminator
+  if (tmp.terminator == '.') {
+    Serial.println("[OK] Terminator valid.");
+  } else {
+    Serial.print("[ERROR] Terminator invalid: '");
+    Serial.print(tmp.terminator);
+    Serial.println("'");
+    return tmp;
+  }
+
+  int datasize = sizeof(tmp.data);
+  if (datasize != 28) {
+    Serial.print("Data of size: "); Serial.println(datasize);
+    if (datasize > 28) {
+      datasize = 28;
+    }
+  }
+
+  // Calculate data CRC
+  uint16_t data_crc = calculateCRC16((uint8_t*) &tmp.data[0], sizeof(datasize));
+
+  // should be only when CRC is good like commented below
+  char inmsg[datasize];
+  memcpy(inmsg, &tmp.data, datasize);
+
+  // Validate incoming data CRC against remote CRC
+  if (tmp.crc16 == data_crc) {
+    Serial.println("[OK] Data CRC valid.");
+    Serial.print("SLAVE Incoming message: "); Serial.println(String(inmsg));
+  } else {
+    Serial.print("CRC-A = 0x");
+    Serial.println(tmp.crc16, HEX);
+    Serial.print("CRC-B = 0x");
+    Serial.println(data_crc, HEX);
+    Serial.print("tmp CRC16: ");
+    Serial.println(tmp.crc16, HEX);
+    Serial.print("SLAVE Incoming message: "); Serial.println(String(inmsg));
+    Serial.println("[ERROR] TODO: Request retransfer exception.");
+    return tmp;
+  }
+
+  // Validate sequence number ( should be already updated from byteParser )
+  uint8_t remote_sequence = tmp.sequence;
+  if (remote_sequence != seq) {
+    Serial.println((char*)tmp.data);
+  }
+
+  msgdata = tmp; // tmp is valid, assign to result address
+
+  return msgdata;
+
+}
+
+void receiveEvent(int howMany) {
+
+  Serial.print("Received "); Serial.print(howMany); Serial.println(" bytes...");
+  char c;
+  char chars[howMany];
+  chars[howMany - 1] = '\0';
+  int index = 0;
+
+  Serial.print("Local Sequence: "); Serial.println((int)seq);
+  digitalWrite(LED_PIN, LOW);
+
+  bool requestRetransfer = false;
+
+  unsigned long rtimeout = millis() + 200; // skip 100ms timeouts
+  while (millis() < rtimeout) {
```

Comment:
How would you solve this without getting involved in something complex? It's not that important, it might be removed instead.
I do understand, that the rtimeout can become a less than millis almost forever in certain moment.

Reply:
I think you should subtract original millis with the current one:

```
  unsigned long rtimeout = millis();
  while (millis() - rtimeout < 200) {
```